cpp/src/traversal/extract_bfs_paths_impl.cuh

```diff
@@ -0,0 +1,378 @@
+      graph_view.get_vertex_partition_last(handle.get_comms().get_size() - 1);
+
+    d_vertex_partition_offsets.resize(handle.get_comms().get_size(), handle.get_stream());
+    rmm::device_uvector<vertex_t> tmp_frontier(current_frontier.size(), handle.get_stream());
+    rmm::device_uvector<vertex_t> original_position(current_frontier.size(), handle.get_stream());
+    rmm::device_uvector<int> original_rank(current_frontier.size(), handle.get_stream());
+
+    raft::update_device(d_vertex_partition_offsets.data(),
+                        h_vertex_partition_offsets.data(),
+                        h_vertex_partition_offsets.size(),
+                        handle.get_stream());
+    raft::copy(
+      tmp_frontier.data(), current_frontier.begin(), current_frontier.size(), handle.get_stream());
+    detail::sequence_fill(
+      handle.get_stream(), original_position.data(), original_position.size(), vertex_t{0});
+    detail::fill(original_rank.size(), original_rank.data(), handle.get_comms().get_rank());
+
+    auto tuple_begin = thrust::make_zip_iterator(
+      thrust::make_tuple(tmp_frontier.begin(), original_position.begin(), original_rank.begin()));
+
+    std::forward_as_tuple(std::tie(tmp_frontier, original_position, original_rank), std::ignore) =
+      groupby_gpuid_and_shuffle_values(
+        handle.get_comms(),
+        tuple_begin,
+        tuple_begin + tmp_frontier.size(),
+        [key_func =
+           detail::mg_partition_vertices<vertex_t>{
+             handle.get_comms().get_rank(),
+             d_vertex_partition_offsets.begin(),
+             d_vertex_partition_offsets.end()}] __device__(auto val) {
+          return key_func(thrust::get<0>(val));
+        },
+        handle.get_stream());
+
+    tuple_begin = thrust::make_zip_iterator(
+      thrust::make_tuple(tmp_frontier.begin(), original_position.begin(), original_rank.begin()));
+
+    thrust::transform(
+      handle.get_thrust_policy(),
+      tuple_begin,
+      tuple_begin + tmp_frontier.size(),
+      tuple_begin,
+      detail::mg_lookup_distance<vertex_t>{graph_view.get_local_vertex_first(), distances});
+
+    std::forward_as_tuple(std::tie(tmp_frontier, original_position, original_rank), std::ignore) =
+      groupby_gpuid_and_shuffle_values(
+        handle.get_comms(),
+        tuple_begin,
+        tuple_begin + tmp_frontier.size(),
+        [] __device__(auto val) { return thrust::get<2>(val); },
+        handle.get_stream());
+
+    thrust::for_each_n(
+      handle.get_thrust_policy(),
+      thrust::make_zip_iterator(
+        thrust::make_tuple(tmp_frontier.begin(), original_position.begin())),
+      tmp_frontier.size(),
+      [max_path_length, d_current_position = current_position.data()] __device__(auto tuple) {
+        auto distance = thrust::get<0>(tuple);
+        auto position = thrust::get<1>(tuple);
+
+        size_t tmp                   = position;
+        d_current_position[position] = (tmp * max_path_length) + distance;
+      });
```

Comment:
So, the code here is basically to collect vertex properties for remote vertices.

I think this will be a recurring pattern, and we may better add something like

collect_values_for_vertices

mimicking the existing collecting_values_for_keys (this assumes that keys are arbitrary rather than contiguous, so using this as is will be somewhat inefficient, but we can copy-and-paste and replace the hash based query with contiguous array accesses with offsets).